plugins/module_utils/vm_snapshot.py

```diff
@@ -0,0 +1,188 @@
+        return dict(
+            uuid=self.uuid,
+            domain_uuid=self.domain_uuid,
+            timestamp=self.timestamp,
+            label=self.label,
+            type=self.type,
+            automated_trigger_timestamp=self.automated_trigger_timestamp,
+            local_retain_until_timestamp=self.local_retain_until_timestamp,
+            remote_retain_until_timestamp=self.remote_retain_until_timestamp,
+            block_count_diff_from_serial_number=self.block_count_diff_from_serial_number,
+            replication=self.replication,
+        )
+
+    # This method is here for testing purposes!
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, VMSnapshot):
+            return NotImplemented
+        return all(
+            (
+                self.uuid == other.uuid,
+                self.domain_uuid == other.domain_uuid,
+                self.timestamp == other.timestamp,
+                self.label == other.label,
+                self.type == other.type,
+                self.automated_trigger_timestamp == other.automated_trigger_timestamp,
+                self.local_retain_until_timestamp == other.local_retain_until_timestamp,
+                self.remote_retain_until_timestamp
+                == other.remote_retain_until_timestamp,
+                self.block_count_diff_from_serial_number
+                == other.block_count_diff_from_serial_number,
+                self.replication == other.replication,
+            )
+        )
+
+    @classmethod
+    def get_by_uuid(
+        cls,
+        ansible_dict: Dict[Any, Any],
+        rest_client: RestClient,
+        must_exist: bool = False,
+    ) -> Optional[VMSnapshot]:
+        query = get_query(ansible_dict, "uuid", ansible_hypercore_map=dict(uuid="uuid"))
+        hypercore_dict = rest_client.get_record(
+            "/rest/v1/VirDomainSnapshot", query, must_exist=must_exist
+        )
+        vm_snapshot_from_hypercore = cls.from_hypercore(hypercore_dict)  # type: ignore
+        return vm_snapshot_from_hypercore
+
+    @classmethod
+    def get_by_snapshots_label(
+        cls,
+        label: str,
+        rest_client: RestClient,
+    ) -> List[Optional[TypedVMSnapshotToAnsible]]:
+        snapshots = [
+            cls.from_hypercore(hypercore_data=hypercore_dict).to_ansible()  # type: ignore
+            for hypercore_dict in rest_client.list_records(
+                "/rest/v1/VirDomainSnapshot", {"label": label}
+            )
+        ]
+
+        return snapshots
+
+    @classmethod
+    def get_by_snapshots_serial(
+            cls,
+            serial: str,
+            rest_client: RestClient,
+    ) -> List[Optional[TypedVMSnapshotToAnsible]]:
+        snapshots = [
+            cls.from_hypercore(hypercore_data=hypercore_dict).to_ansible()  # type: ignore
+            for hypercore_dict in rest_client.list_records(
+                "/rest/v1/VirDomainSnapshot", {"block_count_diff_from_serial_number": serial}
```

Comment:
I think you this will find snapshot just before the desired one.
```
https://10.5.11.50/rest/v1/VirDomainSnapshot/8d7138a0-43c9-4fba-be0a-464c21e7ed02

[
  {
    "uuid": "8d7138a0-43c9-4fba-be0a-464c21e7ed02",
    "domainUUID": "27a86c75-4cb3-4060-84c9-e12d77c7509e",
    "timestamp": 1678602118,
    "label": "clone source XLAB-vm_replication_CI_test",
    "domain": {
      "uuid": "27a86c75-4cb3-4060-84c9-e12d77c7509e",
      "nodeUUID": "",
      "name": "vm-integration-test-source-vm",
      "description": "Source VM",
      "operatingSystem": "os_windows_server_2012",
      "state": "SHUTOFF",
      "desiredDisposition": "SHUTOFF",
      "console": {
        "type": "VNC",
        "ip": "",
        "port": 0,
        "keymap": "en-us"
      },
      "mem": 536870912,
      "numVCPU": 2,
      "blockDevs": [
        {
          "uuid": "9966ed70-cdee-4987-ac68-a5909ead63d7",
          "virDomainUUID": "27a86c75-4cb3-4060-84c9-e12d77c7509e",
          "type": "VIRTIO_DISK",
          "cacheMode": "WRITETHROUGH",
          "capacity": 10844792422,
          "allocation": 0,
          "physical": 0,
          "shareUUID": "",
          "path": "scribe/9966ed70-cdee-4987-ac68-a5909ead63d7",
          "slot": 0,
          "name": "",
          "disableSnapshotting": false,
          "tieringPriorityFactor": 8,
          "mountPoints": [],
          "createdTimestamp": 1675597312,
          "readOnly": false
        }
      ],
      "netDevs": [
        {
          "uuid": "62b16217-72c4-4586-a97e-753ad34ba541",
          "virDomainUUID": "",
          "type": "RTL8139",
          "macAddress": "7C:4C:58:5F:7E:4D",
          "vlan": 1,
          "connected": true,
          "ipv4Addresses": []
        }
      ],
      "stats": [],
      "created": 0,
      "modified": 0,
      "latestTaskTag": {
        "taskTag": "",
        "progressPercent": 0,
        "state": "UNINITIALIZED",
        "formattedDescription": "",
        "descriptionParameters": [],
        "formattedMessage": "",
        "messageParameters": [],
        "objectUUID": "",
        "created": 0,
        "modified": 0,
        "completed": 0,
        "sessionID": "",
        "nodeUUIDs": []
      },
      "tags": "Xlab",
      "bootDevices": [
        "9966ed70-cdee-4987-ac68-a5909ead63d7",
        "62b16217-72c4-4586-a97e-753ad34ba541"
      ],
      "uiState": "SHUTOFF",
      "snapUUIDs": [],
      "snapshotSerialNumber": 119,
      "replicationUUIDs": [],
      "sourceVirDomainUUID": "",
      "snapshotListSerialNumber": 237,
      "snapshotScheduleUUID": "",
      "machineType": "scale-7.2",
      "cpuType": "clusterBaseline-7.3",
      "snapshotAllocationBlocks": 0,
      "guestAgentState": "UNAVAILABLE",
      "lastSeenRunningOnNodeUUID": "",
      "isTransient": false,
      "affinityStrategy": {
        "strictAffinity": false,
        "preferredNodeUUID": "",
        "backupNodeUUID": ""
      },
      "vsdUUIDsToDelete": [],
      "cloudInitData": {
        "userData": "",
        "metaData": ""
      }
    },
    "deviceSnapshots": [
      {
        "uuid": "9966ed70-cdee-4987-ac68-a5909ead63d7",
        "timestamp": 0,
        "serialNumber": 119,
        "blockCountDiff": 0,
        "identifier": "",
        "checksum": ""
      }
    ],
    "unavailable": false,
    "type": "AUTOMATED",
    "blockCountDiff": 0,
    "automatedTriggerTimestamp": 0,
    "localRetainUntilTimestamp": 0,
    "state": "CREATED",
    "quiesced": false,
    "remoteRetainUntilTimestamp": 0,
    "blockCountDiffFromSerialNumber": 118,
    "replication": true
  }
]
```

Partial content:
```
[
  {
    "uuid": "8d7138a0-43c9-4fba-be0a-464c21e7ed02",
    "domainUUID": "27a86c75-4cb3-4060-84c9-e12d77c7509e",
    "timestamp": 1678602118,
    "label": "clone source XLAB-vm_replication_CI_test",
    "domain": {
      "uuid": "27a86c75-4cb3-4060-84c9-e12d77c7509e",
      "nodeUUID": "",
      "name": "vm-integration-test-source-vm",
      "description": "Source VM",
      "operatingSystem": "os_windows_server_2012",
      "state": "SHUTOFF",
      "desiredDisposition": "SHUTOFF",
      "console": {
        "type": "VNC",
        "ip": "",
        "port": 0,
        "keymap": "en-us"
      },
      "mem": 536870912,
      "numVCPU": 2,
      "blockDevs": [
        {
          "uuid": "9966ed70-cdee-4987-ac68-a5909ead63d7",
          "virDomainUUID": "27a86c75-4cb3-4060-84c9-e12d77c7509e",
          "type": "VIRTIO_DISK",
          "cacheMode": "WRITETHROUGH",
          "capacity": 10844792422,
          "allocation": 0,
          "physical": 0,
          "shareUUID": "",
          "path": "scribe/9966ed70-cdee-4987-ac68-a5909ead63d7",
          "slot": 0,
          "name": "",
          "disableSnapshotting": false,
          "tieringPriorityFactor": 8,
          "mountPoints": [],
          "createdTimestamp": 1675597312,
          "readOnly": false
        }
      ],
      "netDevs": [
        {
          "uuid": "62b16217-72c4-4586-a97e-753ad34ba541",
          "virDomainUUID": "",
          "type": "RTL8139",
          "macAddress": "7C:4C:58:5F:7E:4D",
          "vlan": 1,
          "connected": true,
          "ipv4Addresses": []
        }
      ],
      "stats": [],
      "created": 0,
      "modified": 0,
      "latestTaskTag": {
        "taskTag": "",
        "progressPercent": 0,
        "state": "UNINITIALIZED",
        "formattedDescription": "",
        "descriptionParameters": [],
        "formattedMessage": "",
        "messageParameters": [],
        "objectUUID": "",
        "created": 0,
        "modified": 0,
        "completed": 0,
        "sessionID": "",
        "nodeUUIDs": []
      },
      "tags": "Xlab",
      "bootDevices": [
        "9966ed70-cdee-4987-ac68-a5909ead63d7",
        "62b16217-72c4-4586-a97e-753ad34ba541"
      ],
      "uiState": "SHUTOFF",
      "snapUUIDs": [],
      "snapshotSerialNumber": 119,
      "replicationUUIDs": [],
      "sourceVirDomainUUID": "",
      "snapshotListSerialNumber": 237,
      "snapshotScheduleUUID": "",
      "machineType": "scale-7.2",
      "cpuType": "clusterBaseline-7.3",
      "snapshotAllocationBlocks": 0,
      "guestAgentState": "UNAVAILABLE",
      "lastSeenRunningOnNodeUUID": "",
      "isTransient": false,
      "affinityStrategy": {
        "strictAffinity": false,
        "preferredNodeUUID": "",
        "backupNodeUUID": ""
      },
      "vsdUUIDsToDelete": [],
      "cloudInitData": {
        "userData": "",
        "metaData": ""
      }
    },
    "deviceSnapshots": [
      {
        "uuid": "9966ed70-cdee-4987-ac68-a5909ead63d7",
        "timestamp": 0,
        "serialNumber": 119,
        "blockCountDiff": 0,
        "identifier": "",
        "checksum": ""
      }
    ],
    "unavailable": false,
    "type": "AUTOMATED",
    "blockCountDiff": 0,
    "automatedTriggerTimestamp": 0,
    "localRetainUntilTimestamp": 0,
    "state": "CREATED",
    "quiesced": false,
    "remoteRetainUntilTimestamp": 0,
    "blockCountDiffFromSerialNumber": 118,
    "replication": true
  }
]
```

Partial output
```
[
  {
    "uuid": "8d7138a0-43c9-4fba-be0a-464c21e7ed02",
    "domainUUID": "27a86c75-4cb3-4060-84c9-e12d77c7509e",
    "timestamp": 1678602118,
    "label": "clone source XLAB-vm_replication_CI_test",
    "domain": {
      "uuid": "27a86c75-4cb3-4060-84c9-e12d77c7509e",
      "nodeUUID": "",
      "name": "vm-integration-test-source-vm",

      "blockDevs": [
        {
          "uuid": "9966ed70-cdee-4987-ac68-a5909ead63d7",
          "virDomainUUID": "27a86c75-4cb3-4060-84c9-e12d77c7509e",

        }
      ],

      "uiState": "SHUTOFF",
      "snapUUIDs": [],
      "snapshotSerialNumber": 119,
      "sourceVirDomainUUID": "",
      "snapshotListSerialNumber": 237,
    },
    "deviceSnapshots": [
      {
        "uuid": "9966ed70-cdee-4987-ac68-a5909ead63d7",
        "timestamp": 0,
        "serialNumber": 119,
        "blockCountDiff": 0,
        "identifier": "",
        "checksum": ""
      }
    ],
    "unavailable": false,
    "type": "AUTOMATED",
    "blockCountDiff": 0,
    "automatedTriggerTimestamp": 0,
    "localRetainUntilTimestamp": 0,
    "state": "CREATED",
    "quiesced": false,
    "remoteRetainUntilTimestamp": 0,
    "blockCountDiffFromSerialNumber": 118,
  }
]

```

Use `domain.snapshotSerialNumber` ?

Reply:
I left ``get_snapshot_by_query`` alone and instead changed the ``run`` method in ``vm_snapshot_info``.
Now that I think about it again, I might have to move that filtering block of code to method ``get_snapshot_by_query``... we'll see....